master/buildbot/steps/source/git.py

```diff
@@ -267,42 +274,37 @@ def fresh(self):
         yield self._updateSubmodule()
         yield self._cleanSubmodule()
 
+    @defer.inlineCallbacks
     def copy(self):
-        d = self.runRmdir(self.workdir, abandonOnFailure=False)
+        yield self.runRmdir(self.workdir)
 
         old_workdir = self.workdir
         self.workdir = self.srcdir
-        d.addCallback(lambda _: self.incremental())
 
-        def copy(_):
+        try:
+            rc = yield self.incremental()
+            if rc != RC_SUCCESS:
+                raise buildstep.BuildStepFailed()
+
             cmd = buildstep.RemoteCommand('cpdir',
                                           {'fromdir': self.srcdir,
                                            'todir': old_workdir,
                                            'logEnviron': self.logEnviron,
                                            'timeout': self.timeout, })
             cmd.useLog(self.stdio_log, False)
-            d = self.runCommand(cmd)
-            return d
-        d.addCallback(copy)
-
-        def resetWorkdir(_):
+            yield self.runCommand(cmd)
+            if cmd.didFail():
+                raise buildstep.BuildStepFailed()
+            defer.returnValue(RC_SUCCESS)
+        finally:
             self.workdir = old_workdir
-            return RC_SUCCESS
-
-        d.addCallback(resetWorkdir)
-        return d
 
+    @defer.inlineCallbacks
     def finish(self, res):
-        d = defer.succeed(res)
-
-        def _gotResults(results):
-            self.setStatus(self.cmd, results)
-            log.msg("Closing log, sending result of the command %s " %
-                    (self.cmd))
-            return results
-        d.addCallback(_gotResults)
-        d.addCallback(self.finished)
-        return d
+        self.setStatus(self.cmd, res)
+        log.msg("Closing log, sending result of the command %s " %
+                (self.cmd))
+        yield self.finished(res)
```

Comment:
Same point as before.  finished in BuildStep doesn't return anything.  Propagating None manually seems unnecessary since that's the default behaviour for inlineCallback.